### ISSUE TITLE
Error: No known conditions for "./react" entry in "next-auth" package

### ISSUE THREAD

**Issue by Stouffi** [nextauthjs/next-auth#2756]
### Description 🐜

I'm trying to start a server. But at runtime I'm encountering an issue with module resolution.

```
Error: No known conditions for "./react" entry in "next-auth" package
```

I could get it to work by patching the next-auth's package.json tweaking exports field, adding node entries like this:

```json
"exports": {
    ".": {
      "import": "./index.js",
      "node": "./index.js",
    },
    "./jwt": {
      "import": "./jwt/index.js",
      "node": "./jwt/index.js",
    },
    "./react": {
      "import": "./react/index.js",
      "node": "./react/index.js",
    },
    "./providers/*": {
      "import": "./providers/*.js",
      "node": "./providers/*.js",
    }
  },
```

I'm using yarn v3 with PnP and TypeScript v4.4 with module: "CommonJS".

### Is this a bug in your own project?

No

### How to reproduce ☕️

Using yarn v3 with PnP and TypeScript v4.4 with module: "CommonJS".

### Environment 🖥

yarn v3 (canary)
TypeScript 4.4
next-auth@4.0.0-beta.2
CommonJS

### Contributing 🙌🏽

Yes,  I am willing to help solve this bug in a PR

**Comment by balazsorban44** [comment]
I am not particularly fond of yarn v2 and upwards, especially PnP. they do weird stuff, or at least it feels like so. We never got a complaint from others... :confused: 

https://github.com/vercel/next.js/issues/20955 (see all the linked issues at the bottom as well)

Could you test out with yarn v1 for example? if that works, I then have to think it is an upstream issue, not something `next-auth` should be concerned about

**Comment by balazsorban44** [comment]
Looks like @ndom91 had a similar issue, so we might have to look into this. A reproduction is still appreciated, I haven't run into this myself yet. :pray:

**Comment by ndom91** [comment]
Yeah so its only happening on one specific server, where I'm trying to deploy this application, locally I wasn't running into this issue. What I can say about that machine is that it is running Ubuntu 18.04, npm 6 and node 14. I upgraded to npm 7 and node 16 and the issue still appeared when trying to build my project.

Applying the suggested fix above worked for me too in both node 14 + node 16 environments.

**Comment by ndom91** [comment]
So I think we should actually not use the "import" conditional export type in our export map. It explicitly disables require() / cjs / node type importing even though on the server side next-auth may be transpiled to be imported this way in some environments. See: https://nodejs.org/api/packages.html#packages_conditional_exports

Both the example above worked for me, as previously mentioned, but also going back to the "default" export map syntax works for me as well. Like so:

```js
"exports": {
  ".": "./index.js",
  "./jwt":  "./jwt/index.js",
  "./react": "./react/index.js",
  "./providers/*": "./providers/*.js"
},
...
```

**Comment by balazsorban44** [comment]
it was actually changed from that in this commit https://github.com/nextauthjs/next-auth/pull/2552/commits/fcfa07ad0a7578139f1acd2045eb6f859bcb54eb

I used the following guide https://docs.skypack.dev/package-authors/package-checks#check-your-score-before-publishing

**Comment by ndom91** [comment]
Yeah I've used the skypack guide for packages before. They're super aggressive on ESM, which makes sense that they'd recommend going directly for `import`.

**Comment by balazsorban44** [comment]
I feel there should be a middle ground that satisfies their (objectively good) suggestion and the issue at hand. Or is it something we could come back to later on with a better fix? Otherwise maybe your PR is ready and for now we don't care about Skypack scores. 🤷‍♂️

**Comment by Stouffi** [comment]
> So I think we should actually not use the "import" conditional export type in our export map.

Yes this is exactly with what I am in trouble.

After reading more about conditional exports, it's recommended when using "import" condition to also provide "require" (which is mutually exclusive with "import") or "default" which matches any environment otherwise we end up to not be able to resolve those exports when a build is transpiled to CommonJS format and using a modern NodeJS runtime (taking into account exports field). That is also the case with yarn PNP resolver which will strictly follow the conditional exports.

I don't know about Skypack, but as long as we get only one target to use, we could go back with.

```json
"exports": {
  ".": "./index.js",
  "./jwt":  "./jwt/index.js",
  "./react": "./react/index.js",
  "./providers/*": "./providers/*.js"
},
```

Conditional exports are meant to specify different paths like this
```json
"exports": {
  ".": {
    "import": "./dist/es/index.js",
    "require": "./dist/cjs/index.js",
    "default": "./dist/es/index.js"
  },
  "./feature": {
    "import": "./dist/es/feature/index.js",
    "require": "./dist/cjs/feature/index.js",
    "default": "./dist/es/feature/index.js"
  }
},
```

By the way, apart from that issue, a fork of next-auth@beta with a quick fix for exports field, is working very well with yarn@canary with pnp.

**Comment by ndom91** [comment]
Hmm yeah I'd be curious if adding the node/require export map option, in addition to import, would affect the skypack score at all. 

I'm more curious though as to why in my one environment and apparently @Stouffi's, it was transpiled and imported as cjs serverside when that doesn't seem to be the case for most users / environments 🤔 

As a quick fix / workaround I think adding node / require to the export map would be okay. Fred specifically mentions this use case in one section of the docs you linked as well (https://docs.skypack.dev/package-authors/package-checks#how-to-fix-1) so hopefully should affect the score to add node/require in addition to import in the export map (damn what a tongue twister 😅)

**Comment by balazsorban44** [comment]
The thing is, if we added multiple entrypoints, we will have to bundle accordingly. Currently, we only create one type of bundle. Node 12-compatible for server-side modules, and IE11-compatible for the client-side (`next-auth/react`).

I guess ultimately we should create both an IE11 and es6 compatible one for the client, and stick to Node12 and upwards on the server-side.

I am happy with any type of export map that supports this. 😅

**Comment by Stouffi** [comment]
And just to mention it: https://github.com/microsoft/TypeScript/issues/33079

TypeScript currently does not understand exports field, but will soon do it.

For now it just works because the paths of the conditional exports match 1:1 with their target paths.

So waiting for couple of weeks appears to be wise 🙏

**Comment by ndom91** [comment]
Yeah so my use case can definitely wait, I'll just adjust it manually for now. If not many other users mention this then I'd say let's wait for TS and think through a good bundling / exporting strategy.

**Comment by gkielwasser** [comment]
> > So I think we should actually not use the "import" conditional export type in our export map.
> 
> Yes this is exactly with what I am in trouble.
> 
> After reading more about conditional exports, it's recommended when using "import" condition to also provide "require" (which is mutually exclusive with "import") or "default" which matches any environment. Otherwise we end up to not be able to resolve those exports when a build is transpiled to CommonJS format and using a modern NodeJS runtime (taking into account exports field). That is also the case with yarn PNP resolver which will strictly follow the conditional exports.
> 
> I don't know about Skypack, but as long as we get only one target to use, we could go back with.
> 
> ```json
> "exports": {
>   ".": "./index.js",
>   "./jwt":  "./jwt/index.js",
>   "./react": "./react/index.js",
>   "./providers/*": "./providers/*.js"
> },
> ```
> 
> Conditional exports are meant to specify different paths like this
> 
> ```json
> "exports": {
>   ".": {
>     "import": "./dist/es/index.js",
>     "require": "./dist/cjs/index.js",
>     "default": "./dist/es/index.js"
>   },
>   "./feature": {
>     "import": "./dist/es/feature/index.js",
>     "require": "./dist/cjs/feature/index.js",
>     "default": "./dist/es/feature/index.js"
>   }
> },
> ```
> 
> By the way, apart from that issue, a fork of next-auth@beta with a quick fix for exports field, is working very well with yarn@canary with pnp.

I can confirm, i run in a similar issue. I have a shared library that is using client side part code of Nextauth. The library exports commonjs code.
Whenever the transpilled code of the library contains import like ```require('next-auth/react)``` the host application will complain the subpackage is not exported.

For now I have fixed by adding the default case:
```
"exports": {
    ".": {
      "import": "./index.js",
      "default": "./index.js"
    },
    "./jwt": {
      "import": "./jwt/index.js",
      "default": "./jwt/index.js"
    },
    "./react": {
      "import": "./react/index.js",
      "default": "./react/index.js"
    },
    "./providers/*": {
      "import": "./providers/*.js",
      "default": "./providers/*.js"
    }
  }
```

**Comment by ndom91** [comment]
@balazsorban44 okay before the v4 GA release, do we maybe want to change the export map to either:

```js
"exports": {
    ".": {
      "import": "./index.js",
      "default": "./index.js"
    },
    "./jwt": {
      "import": "./jwt/index.js",
      "default": "./jwt/index.js"
    },
    "./react": {
      "import": "./react/index.js",
      "default": "./react/index.js"
    },
    "./providers/*": {
      "import": "./providers/*.js",
      "default": "./providers/*.js"
    }
  }
```

Or the "default" export:
```
"exports": {
    ".": "./index.js",
    "./jwt":  "./jwt/index.js",
    "./react": "./react/index.js",
    "./providers/*": "./providers/*.js"
  }
```

Skypack seems to only car that theres a toplevel export map existent at all in the package.json

**Comment by ndom91** [comment]
Opened a PR #2830 :+1:

**Comment by balazsorban44** [comment]
@Stouffi 

I see we merged #2830, but @ndom91 also opened another PR #2767 which uses your suggestion.

Which one do you think is more correct? :thinking:

**Comment by Stouffi** [comment]
The PR #2830 looks good to me. We should use conditional exports only when we truly need to resolve distinct paths given one subpath pattern.

My first suggestion was based on my poor knowledge about the exports field. But After looking into it in details, it doesn't make sense to use conditional exports if all the conditions lead to the same target.

**Comment by balazsorban44** [comment]
I see. I'll close this issue since #2830 was merged and #2767 then since it seems unnecessary for now. We can revisit the concept when we are able to generate multiple bundles for different targets.
